lib/GridItem.jsx

```diff
@@ -596,22 +582,110 @@ export default class GridItem extends React.Component<Props, State> {
+      handle
     );
 
     // minW should be at least 1 (TODO propTypes validation?)
     minW = Math.max(minW, 1);
 
-    // maxW should be at most (cols - x)
-    maxW = Math.min(maxW, cols - x);
-
     // Min/max capping
     w = clamp(w, minW, maxW);
     h = clamp(h, minH, maxH);
 
-    this.setState({ resizing: handlerName === "onResizeStop" ? null : size });
+    let updatedSize = size;
+    if (node) {
+      const currentLeft = position.left;
+      const currentTop = position.top;
+      const currentWidth = position.width;
+      const currentHeight = position.height;
+
+      const constrainWidth = width => ({
+        width:
+          currentLeft + width > this.props.containerWidth ? currentWidth : width
+      });
+
+      const constrainHeight = (top, height) => ({
+        height: top <= 0 ? currentHeight : height
+      });
+
+      const constrainLeft = left => ({
+        left: Math.max(0, left)
+      });
+
+      const constrainTop = top => ({
+        top: Math.max(0, top)
+      });
+
+      const resizeNorth = ({ left, height, width }) => {
+        const top = currentTop - (height - currentHeight);
+
+        return {
+          top,
+          left,
+          width,
+          ...constrainHeight(top, height),
+          ...constrainTop(top)
+        };
+      };
+
+      const resizeEast = ({ top, left, height, width }) => ({
+        top,
+        height,
+        ...constrainWidth(width),
+        ...constrainLeft(left)
+      });
+
+      const resizeWest = ({ top, height, width }) => {
+        const left = currentLeft - (width - currentWidth);
+
+        return {
+          height,
+          width: left <= 0 ? currentWidth : constrainWidth(width).width,
+          ...constrainTop(top),
+          ...constrainLeft(left)
+        };
+      };
+
+      const resizeSouth = ({ top, left, height, width }) => ({
+        width,
+        left,
+        ...constrainHeight(top, height),
+        ...constrainTop(top)
+      });
+
+      const resizeNorthEast = (...args) => resizeNorth(resizeEast(...args));
+      const resizeNorthWest = (...args) => resizeNorth(resizeWest(...args));
+      const resizeSouthEast = (...args) => resizeSouth(resizeEast(...args));
+      const resizeSouthWest = (...args) => resizeSouth(resizeWest(...args));
+
+      const ordinalResizeHandlerMap = {
+        n: resizeNorth,
+        ne: resizeNorthEast,
+        e: resizeEast,
+        se: resizeSouthEast,
+        s: resizeSouth,
+        sw: resizeSouthWest,
+        w: resizeWest,
+        nw: resizeNorthWest
+      };
+
+      const resizeHandler = ordinalResizeHandlerMap[handle];
+      updatedSize = resizeHandler
+        ? resizeHandler({
+            top: currentTop,
+            left: currentLeft,
+            height: currentHeight,
+            width: currentWidth,
+            ...size
+          })
+        : size;
+      this.setState({
+        resizing: handlerName === "onResizeStop" ? null : updatedSize
+      });
+    }
```

Comment:
This entire block of changes is **much** more verbose than before, but I think its a lot more readable, maintainable, and extensible now.